Use EF for SQLite interfacing

UselessToucan opened the pull request:
Use EF Core for database interaction.

Depends on [this](https://github.com/ppy/osu-framework/pull/1075) framework PR.

UselessToucan committed:
Initial EF Core commit

UselessToucan committed:
Rework EF Core usage in model classes

UselessToucan committed:
Use one DbContext with configured CommandTimeout for all stores

UselessToucan committed:
Fix foreign key constraint failure

UselessToucan committed:
Add missing eager loading

UselessToucan committed:
Small DB-related refactoring

UselessToucan committed:
Fix BeatmapMetadata removing

UselessToucan committed:
Merge from ppy/master

UselessToucan committed:
Fix RulesetStore preparation

peppy commented:
Why all the renames though?

peppy commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -168,7 +168,7 @@ public BeatmapSetInfo Import(ArchiveReader archiveReader)
 
             // let's only allow one concurrent import at a time for now.
             lock (importLock)
-                connection.RunInTransaction(() => Import(set = importToStorage(archiveReader)));
```
what happened to transaction logic?

UselessToucan replied:
AFAIK SaveChanges method uses transactions under the hood. Also, it is possible to use transactions manually using DbContext.Database.BeginTransaction() method.
https://docs.microsoft.com/en-us/ef/core/saving/transactions

peppy replied:
Sure, but from what I can see here you removed the transaction and didn’t leave a todo or anything.

peppy commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -502,16 +492,12 @@ private BeatmapSetInfo importToStorage(ArchiveReader reader)
-        /// <param name="populate">Whether returned objects should be pre-populated with all data.</param>
         /// <returns>A list of available <see cref="BeatmapSetInfo"/>.</returns>
-        public List<BeatmapSetInfo> GetAllUsableBeatmapSets(bool populate = true)
+        public List<BeatmapSetInfo> GetAllUsableBeatmapSets()
```
why removing functionality?

UselessToucan replied:
Eager loading is implemented in Query methods(https://github.com/UselessToucan/osu/blob/EfCore/osu.Game/Beatmaps/BeatmapStore.cs#L138).
If this is critical for you, I will try to extract it to Populate method.

peppy commented on osu.Game/Database/DatabaseBackedStore.cs:
```diff
@@ -2,27 +2,23 @@
-        protected readonly SQLiteConnection Connection;
+        protected readonly OsuDbContext Connection;
 
-        protected virtual int StoreVersion => 1;
```
????

UselessToucan replied:
As far as I understand, this property was used only for DB migrations. In EF we can migrate DB without it. I can just call `dbContext.Database.Migrate();` and pending migrations will be applied(if there are any).

peppy commented on osu.Game/Database/DbContextBase.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace osu.Game.Database
+{
+    public abstract class DbContextBase:DbContext
```
what is this

peppy commented on osu.Game/Database/OsuDbContext.cs:
```diff
@@ -0,0 +1,49 @@
+    {
+        private readonly string connectionString;
+
+        public OsuDbContext()
```
what is this? why are there no comments?

UselessToucan replied:
My bad. Will add comments this evening when I get back home.

peppy commented on osu.Game/Database/OsuDbContext.cs:
```diff
@@ -0,0 +1,49 @@
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            base.OnModelCreating(modelBuilder);
+            modelBuilder.Entity<BeatmapInfo>().HasIndex(b => b.MD5Hash);
```
no

peppy commented on osu.Game/OsuGameBase.cs:
```diff
@@ -82,35 +83,41 @@ public OsuGameBase()
-            conn.BusyTimeout = new TimeSpan(TimeSpan.TicksPerSecond * 10);
-            return conn;
+            var connectionString = Host.Storage.GetDatabaseConnectionString(@"client");
+            var context = new OsuDbContext(connectionString);
```
this is all stupid. why is this not at the framework level? why is sqlite implemented in osu.game? why is there even anything left in the framework? it's pointless at this point because all the implementation (except for the connection string?) is here.

and it's in OsuGameBase

no please god no

UselessToucan replied:
I can't create OsuDbContext at the framework level because it must contain DbSets(which are game specific).

peppy replied:
if t can’t be abstracted to the framework then it shouldn’t exist in the framework, but i’m pretty sure it can be.

peppy commented:
Please begin by undoing the naming changes and keeping this PR to the minimal required implementation.

UselessToucan commented:
I reviewed naming changes and noticed that most of them are Id properties.
It is [EF Convention](https://docs.microsoft.com/en-us/ef/core/modeling/relationships#conventions) for primary/foreign keys.

smoogipoo commented:
I would rather us temporarily not abiding to conventions if it means not going through a 2000 line change PR.

peppy commented:
Also what are the .Designer files generated from? We'll probably not want to do that, and generate by hand to fix bad formatting and allow people who aren't using (whatever IDE created them) to still work on the project easily.

UselessToucan commented:
> Also what are the .Designer files generated from?
20171009205048_Init.cs, 20171009205048_Init.Designer.cs, OsuDbContextModelSnapshot.cs were generated by "Add-Migration <Name>" call. These files are required for creating(or updating) database on the client machine.

peppy commented on osu.Game/Beatmaps/BeatmapSetOnlineInfo.cs:
```diff
@@ -12,11 +15,19 @@ namespace osu.Game.Beatmaps
     public class BeatmapSetOnlineInfo
     {
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public int Id { get; set; }
```
why is this added (and why is this class modified)? currently this stuff isn't stored in the database

UselessToucan replied:
Yes, there is actually no need to modify this class. It seems that I overdid it.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into UselessToucan/EfCore

# Conflicts:
#	osu-framework
#	osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs
#	osu.Game/Beatmaps/BeatmapInfo.cs
#	osu.Game/Online/API/Requests/GetBeatmapSetsResponse.cs

smoogipoo committed:
Fix compile errors

smoogipoo committed:
Let's not rename columns yet

smoogipoo committed:
Add logging

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into UselessToucan/EfCore

# Conflicts:
#	osu-framework
#	osu.Game/Beatmaps/BeatmapMetadata.cs
#	osu.Game/osu.Game.csproj
#	osu.Game/packages.config

smoogipoo committed:
Update framework

peppy committed:
Initialise batteries

Note that this is in the wrong place. But so is the rest of this code.

UselessToucan committed:
Add System.ValueTuple NuGet package

UselessToucan committed:
Return back DatabaseBackedStore's query and populate functions

huoyaoyuan committed:
Construct DwarableScore using null weight.

huoyaoyuan committed:
Use localisation engine instead of asking current culture directly.

huoyaoyuan committed:
Use format string for double instead of Math.Round.

huoyaoyuan committed:
CI fixes.

UselessToucan committed:
Fix removal of FileInfo, BeatmapMetadata, BeatmapDifficulty objects

peppy committed:
Update framework

peppy committed:
Fix some file headers and unused files

peppy committed:
Add one more licence header

peppy committed:
Update CodeFileSanity to ignore migrations folder

peppy committed:
Update appveyor worker image to vs2017

peppy committed:
Ignore migrations directory for inspections

UselessToucan committed:
Fix inconsistent lock usage in BeatmapManager

UselessToucan committed:
Revert "Fix inconsistent lock usage in BeatmapManager"

This reverts commit 4a064da30f6349a7d2f327aa2b2769635fe02037.

peppy committed:
Tidy up context creation

peppy committed:
Update framework

peppy committed:
Not implemented instead of incorrect

peppy committed:
Revert "Fix removal of FileInfo, BeatmapMetadata, BeatmapDifficulty objects"

peppy committed:
Revert "Return back DatabaseBackedStore's query and populate functions"

This reverts commit 7cf5d63cd32bde05ee0bcbc262e5dac7595036f4.

peppy committed:
Fix test regression

peppy committed:
Fix remaining warnings

peppy committed:
Fix formatting regressions

peppy committed:
Add missing reference to System.ValueTuple

peppy committed:
Fix ValueTuple mess

peppy committed:
Fix all keybindings being reset every startup

peppy committed:
Add BindingRedirects rule to fix dependency when running via nunit

See https://github.com/ErikEJ/EntityFramework.SqlServerCompact/issues/463

peppy committed:
Remove ValidTypes from databased stored; explicitly expose query methods instead

peppy committed:
Fix test case runs not being correctly isolated on mono

peppy committed:
Remove duplicate command

peppy committed:
Correct and simplify RulesetStore

peppy committed:
Correct and simplify KeyBindingStore

peppy committed:
Merge remote-tracking branch 'upstream/master' into EfCore

peppy commented on osu.Game/osu.Game.csproj:
```diff
@@ -116,27 +160,52 @@
-    <Reference Include="SQLiteNetExtensions, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>$(SolutionDir)\packages\SQLiteNetExtensions.1.3.0\lib\portable-net45+netcore45+wpa81+wp8+MonoAndroid1+MonoTouch1\SQLiteNetExtensions.dll</HintPath>
+    <Reference Include="System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51">
+      <HintPath>../packages/System.ValueTuple.4.4.0/lib/net461/System.ValueTuple.dll</HintPath>
```
/ \

peppy committed:
Remove weird additions

peppy committed:
CI fixing

peppy committed:
Correct and simplify BeatmapStore and BeatmapManager

UselessToucan commented on osu.Game/Beatmaps/BeatmapStore.cs:
```diff
@@ -125,46 +127,20 @@ public bool Restore(BeatmapInfo beatmap)
-                .Include(b => b.Difficulty)
-                .Where(query).ToList();
-        }
+        public IEnumerable<BeatmapSetInfo> BeatmapSets => Connection.BeatmapSetInfo
```
I like these two properties.
Shame on me because I did not come to this implementation from the very beginning.

peppy commented:
remaining tasks are

- go through the actual EF database structure, including fixing the cascade deletion to FileInfo and BeatmapDifficulty which may or may not be broken
- reinstate transactions (i've already read into it and it doesn't look to be an issue)
- comprehensive testing of multi-connection interaction (since the new code is switching to provide a db context per store)
- fixing of the fucked up ValueTuple crap which is causing CI tests to fail (this is a really shocking situation, there's hundreds of people having the same issue due to weird versioning across different .net platforms)
- adding more tests for each store type
- considering migration from old db format

peppy committed:
FileStore logic fixes

peppy committed:
osu.Game csproj fixes

peppy committed:
Fix intro and duplicate inserts

peppy committed:
Simplify file storing

peppy committed:
Fix username display on beatmap panels

peppy committed:
Connection -> context

peppy committed:
Add back startup tasks runner

peppy committed:
Add back transaction support for beatmap importing

peppy committed:
Standardise context retrieval

peppy committed:
Fix not being able to restore hidden beatmaps via context menu

UselessToucan commented:
> fixing the cascade deletion to FileInfo and BeatmapDifficulty which may or may not be broken

@peppy [You reverted this one yourself](https://github.com/ppy/osu/pull/1350/commits/04e5f764a34e807db1a25de355bd79709f3aac74) without explaining what is wrong.

peppy commented:
i already said i reverted *all your changes*. also that one was *explicitly incorrect* because the foreign key was in the wrong table.

peppy committed:
Fix TestCasePlaySongSelect

peppy committed:
Fix KeyBindingStore regression

peppy committed:
Fix file prefix not being read when calling storage.Exists

peppy committed:
Improve performance of beatmap imports (still needs revision)

peppy committed:
Fix xmldoc

peppy committed:
Fix tooling failures

peppy committed:
Use a different database name for now to avoid conflicts when switching versions

peppy committed:
Fix cascade deletions

peppy committed:
Run cleanup tasks only on startup via manual calls

peppy committed:
Add some more missing indices

peppy committed:
Improve performance of imports by keeping a context hot

peppy committed:
Update framework

peppy committed:
Merge remote-tracking branch 'upstream/master' into EfCore

peppy committed:
Remove unnecessary locking; operations are now thread-safe

peppy committed:
Fix deletion and use single context for imports and deletions for now

peppy committed:
Remove some unnecessary update calls

peppy committed:
Fix beatmap carousel interactions with deletion when not yet displayed

peppy committed:
Ignore r# locked context inspection here

peppy committed:
Remove TestCase cleanup temporarily until context disposal is sorted

peppy committed:
valuetuple nightmare

peppy committed:
Query test assert conditions less often

peppy committed:
FIx missing columns yet again

peppy committed:
Fix all remaining db structure issues

peppy committed:
ValueTuple please

peppy committed:
Update app config for rulesets

smoogipoo commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -156,21 +171,39 @@ public void Import(params string[] paths)
+            {
+                var context = importContext.Value;
+
+                context.Database.AutoTransactionsEnabled = false;
```
Is this needed? Why is this not needed inside Delete()?

peppy replied:
to be honest i'm not sure this has much of a benefit; requires further testing. basically it runs `DetectChanges` less than otherwise would be. When I get to addressing transactions via a better pattern it will likely be automatically toggled if there are benefits.

If you'd prefer we can remove it for now. Or add it to deletion.

smoogipoo replied:
Also why is this not being reset back to true?

peppy replied:
oh, ignore previous comment, i misread which bool this was. it needs to be set for deletions too.

peppy committed:
Remove unnecessary second call to AutoTransactionsEnabled

smoogipoo reviewed (dismissed):
This looks pretty good. I've got not major issues with it, beyond that it needs to be abstracted super hard, and needs to be moved to the framework.

peppy committed:
Reduce database log output

Also hard-disables it for uninteresting log levels, providing a further performance boost.

peppy committed:
Don't make one factory each context

A factory is supposed to be re-used.

peppy committed:
Fix beatmap difficulty and metadata deletion

peppy committed:
Allow migration from sqlite-net to EF

Bonus stage

smoogipoo commented on osu.Game/Database/DatabaseContextFactory.cs:
```diff
@@ -14,6 +14,6 @@ public DatabaseContextFactory(GameHost host)
         }
 
-        public OsuDbContext GetContext() => new OsuDbContext(host.Storage.GetDatabaseConnectionString(@"client-ef"));
+        public OsuDbContext GetContext() => new OsuDbContext(host.Storage.GetDatabaseConnectionString(@"client"));
```
So what happens when I run TestCasePlaySongSelect?

peppy replied:
Nothing has changed there; it uses a facade storage which resolves to an in-memory SQLite instance

smoogipoo commented on osu.Game/Database/OsuDbContext.cs:
```diff
@@ -156,5 +156,74 @@ public bool IsEnabled(LogLevel logLevel)
+                    Database.ExecuteSqlCommand("DROP TABLE BeatmapInfo_Old");
+                }
+                catch
+                {
```
Is this going to be a problem if this fails on an older DB?

peppy committed:
Remove AllRuleset references

peppy committed:
Add fallback logic in case migration fails

Nuke it all.

peppy committed:
Merge branch 'master' into EfCore

peppy committed:
Add fallback logic for the case where previous database can't be migrated
